### PR TITLE
Enhancement: Add support for ENS lookup and simple registration

### PR DESCRIPTION
**Pull request by eggplantzzz**
This PR adds support for ENS names. Users will be able to enter an ENS name where they can enter addresses and Truffle will resolve the address.

It also aims to support simple registration of ENS names. A user will be able to use a syntax like ~`deployer.ens.register('myName.eth')`~`deployer.ens.setAddress('myName.eth')`. Truffle will check to see if the name is owned by the address sending the contract. If so it will attempt to set the address for that resolver for that name. If the name is owned by another address it will fail.

### CODE REVIEW

**Comment by eggplantzzz** [comment]
There might be an issue with automatic name registration. Currently I do not believe there are any standard "interface types" which indicate the ability to automatically register a name. There is [EIP165](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-165.md) but there are currently no established official interface types.

**Comment by eggplantzzz** [comment]
It was noted by @gnidan that some of the libraries included in truffle-deployer (and possibly truffle-resolver) might need to be included in truffle to make sure they make their way into the bundle. Among these might be the ens contract dependencies (@ensdomains/ens | @ensdomains/resolvers).

File: packages/deployer/ens.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+      let registryAddress = this.determineENSRegistryAddress();
+
+      publicResolver = await PublicResolver.new(registryAddress, { from });
+      await this.ensjs.setResolver(name, publicResolver.address, { from });
+      return { resolvedAddress: null };
+    }
+  }
+
+  async setAddress(name, addressOrContract, from) {
```
Comment (eggplantzzz) [inline packages/deployer/ens.js]: `from` here should probably be a tx object instead like normal deployer syntax

File: packages/deployer/index.js [hunk 2]
```diff
@@ -13,6 +14,8 @@ class Deployer extends Deployment {
     const emitter = new Emittery();
     super(emitter, options);
 
+    const { ens } = options.networks[options.network] || {};
```
Comment (eggplantzzz) [inline packages/deployer/index.js]: Add note somewhere here that ens is configured globally as well as per network.

**Comment by haltman-at** [comment]
Repeating comment from meeting: I notice that when handling ENS names given as arguments, it only handles them if they're done at top level, not if they're inside arrays or structs.  I mean, we seem to have agreed that that's fine for now, but I just kind of wanted to record it here for the future.

**Comment by haltman-at** [comment]
Oh right and the other comment from the meeting: I think resolution (or at least resolution; possibly more) should probably be in a separate package from `contract`, because I think `codec` is going to want to use that functionality later.  If `contract` also ends up using `codec` later as I'm planning, this would result in circular dependencies.

(Mind you there might be ways to do it without this problem, e.g. @gnidan suggested a generator-based solution (which I'm not sure I like but there are likely other similar solutions), so this may be premature, but I figured just in case, to avoid potentially breaking things later...)

**Comment by gnidan** [comment]
Docs PR: https://github.com/trufflesuite/trufflesuite.com/pull/498

File: packages/contract/lib/utils/index.js [hunk 3]
```diff
@@ -59,12 +60,7 @@ const Utils = {
           copy.data = "";
         }
 
-        let logArgs;
-        try {
-          logArgs = abi.decodeLog(logABI.inputs, copy.data, copy.topics);
-        } catch (_) {
-          return null;
-        }
+        const logArgs = abi.decodeLog(logABI.inputs, copy.data, copy.topics);
```
[the remark this answers is not in the document]
Comment (eggplantzzz) [reply packages/contract/lib/utils/index.js]: Hmm, I don't remember why I removed it. I guess we should keep it in.

File: packages/provisioner/index.js [hunk 4]
```diff
@@ -13,6 +13,8 @@ var provision = function(abstraction, options) {
     abstraction.setNetworkType(options.networks[options.network].type);
   }
 
+  abstraction.ens = options.ens;
```
[the remark this answers is not in the document]
Comment (eggplantzzz) [reply packages/provisioner/index.js]: I suppose there is a pattern here :)

[the remark this answers is not in the document]
Comment (eggplantzzz) [reply packages/provisioner/index.js]: Well actually we still might want the default values there. Let me look into it a bit.

[the remark this answers is not in the document]
Comment (CruzMolina) [reply packages/provisioner/index.js]: Just thinking for non-eth contract instances. Not really sure if it matters.

**Pull request merged**